In [421]:
import pandas as pd
import json
from datetime import datetime

In [377]:
df = pd.read_csv('hackathon_data.csv')

/var/folders/j5/_zyfcd891fd2sh3fg54z3gmh0000gn/T/ipykernel_84639/1649436523.py:1: DtypeWarning: Columns (11,12) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('hackathon_data.csv')


In [378]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 13 columns):
 #   Column              Non-Null Count    Dtype  
---  ------              --------------    -----  
 0   id                  1000000 non-null  int64  
 1   Marka               1000000 non-null  object 
 2   IsyeriTipi          1000000 non-null  object 
 3   UyeDurum            1000000 non-null  object 
 4   UyeAdres            1000000 non-null  object 
 5   Sehir               1000000 non-null  object 
 6   UyeAktivasyonTarih  1000000 non-null  object 
 7   IslemID             1000000 non-null  int64  
 8   IslemTarih          1000000 non-null  object 
 9   IslemTutar          1000000 non-null  float64
 10  Tercih              1000000 non-null  object 
 11  Versiyon            23643 non-null    object 
 12  SurumTarih          23643 non-null    object 
dtypes: float64(1), int64(2), object(10)
memory usage: 99.2+ MB


In [379]:
df['UyeAdres'] = df['UyeAdres'].str.upper()

In [380]:
 df['Sehir'].unique()


array(['Samsun', 'İstanbul', 'Manisa', 'İzmir', 'Aksaray', 'Antalya',
       'Bursa', 'Kocaeli', 'Ankara', 'Eskişehir', 'Giresun', 'Gaziantep'],
      dtype=object)

In [381]:
df.loc[df['UyeAdres'] == 'DEREBOYU 1', 'UyeAdres'] = 'DEREBOYU CD. NO:1 MECIDIYE, 34347 BEŞIKTAŞ/ İSTANBUL'

In [382]:
df.loc[df['UyeAdres'] == 'DEREBOYU CAD NO 1', 'UyeAdres'] = 'DEREBOYU CD. NO:1 MECIDIYE, 34347 BEŞIKTAŞ/ İSTANBUL'

In [383]:
df.loc[df['UyeAdres'] == 'KARTAL', 'UyeAdres'] = 'KARTAL/ İSTANBUL'

In [384]:
df.loc[df['UyeAdres'] == 'ATAŞEHIR TEST ADRESI', 'UyeAdres'] = 'ATAŞEHİR/ İSTANBUL'

In [385]:
df.loc[df['UyeAdres'] == 'BÜYÜKSARAY MEYDANI CAD NO:42', 'UyeAdres'] = 'BÜYÜK SARAY MEYDANI CD. NO:42 ŞEHREMINI, 34104 FATIH/ İSTANBUL'

In [386]:
df.loc[df['UyeAdres'] == 'BAŞARAN CAFE RESTORAN OTOMOTİV VE TARIM ANONİM ŞİRKETİ', 'UyeAdres'] = 'GÜZELYURT MAH. ORHANGAZI CAD. A2 APT. NO:5 BI ESENYURT/ İSTANBUL'

In [387]:
df.loc[df['UyeAdres'] == 'ISKENDER PAŞA MAH. SOFULAR TEKKESI SOKAK NO 18 A', 'UyeAdres'] = 'SOFULAR TEKKESI SOK., NO:18A, İSKENDERPAŞA MAH., FATIH/ İSTANBUL.'

In [388]:
df.loc[df['UyeAdres'] == 'YEŞILCE MAH. DESTEGÜL SOK NO 1 A', 'UyeAdres'] = 'YEŞILCE MAH. DESTEGÜL SOK. 1/A KAĞITHANE/ İSTANBUL'

In [389]:
df.loc[df['UyeAdres'] == 'YENİKENT MAH. İSMETPAŞA CAD. ÇAMLIK SİTESİ A BİLOK NO 51 4/7 DERİNCE/KOCAELİ', 'UyeAdres'] = 'YENİKENT MAH. İSMETPAŞA CAD. ÇAMLIK SİTESİ A BİLOK NO 51 4/7 DERİNCE/ KOCAELİ'


In [390]:
unique_adresler = df['UyeAdres'].unique()
unique_adresler

array(['KALE MAH. GAZİ CAD. 82 C İLKADIM/ SAMSUN',
       'DEREBOYU CD. NO:1 MECIDIYE, 34347 BEŞIKTAŞ/ İSTANBUL',
       'KOCATEPE MAH. ŞEHİR PARKI CAD NO: 105 İÇ KAPI NO: 107 BAYRAMPAŞA/ İSTANBUL',
       'ARMAĞANEVLER MAH. HARMAN SK. NO: 17 İÇ KAPI NO: 17 ÜMRANİYE/ İSTANBUL',
       'KURTULUŞ MAH. HÜRRİYET BLV. NO: 11 A İÇ KAPI NO: A SOMA/ MANİSA',
       'VALİ RAHMİ BEY MAH. MENDERES CAD. AKABE NO: 55 İÇ KAPI NO: 1 BUCA/ İZMİR',
       'ZAFER MAH/SEMT KIRŞEHİR CAD. 185 ORTAKÖY/ AKSARAY',
       'ERZENE MAH. KAZIM KARABEKİR CAD. NO: 4 G BORNOVA/ İZMİR',
       'ALTINOVA SİNAN MAH. ÇAĞDAŞ SK. AGORA AVM - NO: 5 KEPEZ/ ANTALYA',
       'ZAFER MAH. ÇINAR 1 SK. 5 B BAHÇELİEVLER/ İSTANBUL',
       'KARTAL/ İSTANBUL',
       'ATAKENT MAH. MUAMMER AKSOY SK. NO: 6 A İÇ KAPI NO: A ÜMRANİYE/ İSTANBUL',
       'HAMİTLER MAH. 1.ZEKİ SK. E BLOK NO: 10 A OSMANGAZİ/ BURSA',
       'SEYRANTEPE MAH. AKSEV SK. 2 2 KAĞITHANE/ İSTANBUL',
       'YEŞILCE MAH. DESTEGÜL SOK. 1/A KAĞITHANE/ İSTANBUL',
      

In [391]:
unique_df = pd.DataFrame(unique_adresler, columns=['UyeAdres'])

In [392]:
unique_df.to_csv("benzersiz_adresler.csv", index=False)

In [393]:
df['Sehir']= [''.join(str(i).split(' ')[-2:]) for i in df['UyeAdres']]


#### Verileri Düzenledik. Şimdi Üzerinde Ne Analizler Yapabiliriz Bakıyoruz..

In [394]:
df.to_csv('new_dataframe.csv')

In [395]:
df.sample(10)

id          Marka     IsyeriTipi UyeDurum  \
371551  301025172  Biye bakaliye  Şahıs Şirketi    Aktif   
767501  301029860  Oden o- maden   Tüzel Şirket    Pasif   
561698  301026102  Kıda kaz gıda  Şahıs Şirketi    Aktif   
156320  301012873  Eşti ekd. şti   Tüzel Şirket    Aktif   
800499  301030448  Alik arncilik   Tüzel Şirket    Pasif   
920052  301032425  Aziz ab leziz  Şahıs Şirketi    Aktif   
878910  301031339  Bant baaurant   Tüzel Şirket    Aktif   
393205  301025608  Nfka nu yufka  Şahıs Şirketi    Aktif   
266261  301021277  Rimi renetimi   Tüzel Şirket    Aktif   
631148  301026648  Sesi suzanesi  Şahıs Şirketi    Pasif   

                                                 UyeAdres  \
371551  HIRKA-İ ŞERİF MAH. YAMAK SK. NO: 4 İÇ KAPI NO:...   
767501  ESENTEPE MAH. KORE ŞEHİTLERİ CAD. 4 -6 20 ŞİŞL...   
561698   GÜLTEPE MAH. SERHAT SK. 9 KÜÇÜKÇEKMECE/ İSTANBUL   
156320  ERZENE MAH. KAZIM KARABEKİR CAD. NO: 4 G BORNO...   
800499  KEMALPAŞA MAH. HALKALI CAD. ATA İŞ MERKEZİ 103...   
920052   BULANCAK MAH. ADLİYE SK. NO: 4 BULANCAK/ GİRESUN   
878910  GÜZELYURT MAH. ORHANGAZI CAD. A2 APT. NO:5 BI ...   
393205  MERKEZ MAH. Ş.YILMAZ ÖZDEMİR CAD. 61 4 KÜÇÜKÇE...   
266261  SEYRANTEPE MAH. AKSEV SK. 2 2 KAĞITHANE/ İSTANBUL   
631148  KEMALPAŞA MAH BAĞLAR CAD 18 B KÜÇÜKÇEKMECE/ İS...   

                        Sehir   UyeAktivasyonTarih    IslemID  \
371551         FATİH/İSTANBUL  2018-01-10 15:40:18  501256817   
767501         ŞİŞLİ/İSTANBUL  2018-09-14 10:19:24  404786136   
561698  KÜÇÜKÇEKMECE/İSTANBUL  2018-02-21 18:36:14  546603402   
156320          BORNOVA/İZMİR  2022-08-16 08:48:29  403550826   
800499  KÜÇÜKÇEKMECE/İSTANBUL  2018-10-23 19:05:38  470054282   
920052       BULANCAK/GİRESUN  2018-12-28 15:13:16  443833989   
878910      ESENYURT/İSTANBUL  2018-11-24 17:41:17  406607374   
393205  KÜÇÜKÇEKMECE/İSTANBUL  2018-01-26 15:08:50  403274403   
266261     KAĞITHANE/İSTANBUL  2017-07-19 10:24:28  411098803   
631148  KÜÇÜKÇEKMECE/İSTANBUL  2018-03-13 17:40:36  408805625   

                 IslemTarih  IslemTutar    Tercih Versiyon SurumTarih  
371551  2023-01-23 23:25:00       17.00       ÖKC      NaN        NaN  
767501  2019-10-23 10:04:56       30.00       ÖKC      NaN        NaN  
561698  2023-06-08 19:08:00       70.00       ÖKC      NaN        NaN  
156320  2019-04-19 07:03:00       15.75       ÖKC      NaN        NaN  
800499  2022-08-19 23:47:00      250.00       ÖKC      NaN        NaN  
920052  2022-01-31 13:30:00       72.00       ÖKC      NaN        NaN  
878910  2020-02-02 05:54:00      140.50       ÖKC      NaN        NaN  
393205  2019-02-08 19:37:00       21.50       ÖKC      NaN        NaN  
266261  2020-09-10 09:19:25      810.00  CeptePos      NaN        NaN  
631148  2020-06-12 16:13:00      149.00       ÖKC      NaN        NaN

#### Markaların isimleri bize anlamlı bir mesaj verebilir mi?

In [449]:
df['Marka'].unique()

array(['Kzar kü pazar', 'Teri tejileri', 'Aida aka gida', 'Hdeo he video',
       'Hiye hatasiye', 'Hkek haan kek', 'Wtma waarıtma', 'Ayon arvasyon',
       'Eşti ekd. şti', 'Took tatawook', 'Wsky woversky', 'Ka.ş kiva a.ş',
       'Bant boaurant', 'Asua axaxesua', 'Rimi renetimi', 'Öeti ödirketi',
       'Kayi kol bayi', 'Aüfe ass büfe', 'Biye bakaliye', 'Llım leazılım',
       'Nfka nu yufka', 'Kıda kaz gıda', 'Küfe kar büfe', 'Hs 3 hy sis 3',
       'Aket akmarket', 'Sesi suzanesi', 'Eket ermarket', 'Ezgi ezezgi',
       'Bkal babakkal', 'Örek öz börek', 'Oden o- maden', 'Giye göhafiye',
       'Alik arncilik', 'Mand mad land', 'Blas burillas', 'Bant baaurant',
       'Aziz ab leziz', 'Dkli doymakli', 'Öket özmarket'], dtype=object)

#### Marka Tarih kırılımında hangi marka hangi tarih aralıklarında işlem yapmış bi inceleyelim

In [450]:
df['IslemTarih'] = pd.to_datetime(df['IslemTarih'])

grouped = df.groupby(['id', 'Marka'])['IslemTarih'].agg(['min', 'max']).reset_index()

print(grouped)


           id          Marka                 min                 max
0   301002470  Kzar kü pazar 2015-09-18 14:35:04 2023-10-26 10:47:00
1   301002583  Teri tejileri 2014-09-25 10:08:52 2020-12-23 12:28:28
2   301003354  Aida aka gida 2022-11-21 18:05:00 2023-10-26 16:59:00
3   301006906  Hdeo he video 2016-01-29 11:00:43 2022-09-12 18:30:00
4   301009366  Hiye hatasiye 2019-06-27 17:31:14 2022-09-19 18:01:00
5   301009412  Hkek haan kek 2016-08-26 11:00:35 2023-10-03 17:56:28
6   301009682  Wtma waarıtma 2016-05-19 10:29:01 2023-10-26 16:39:00
7   301010389  Ayon arvasyon 2016-06-15 15:12:25 2020-05-10 11:09:14
8   301012873  Eşti ekd. şti 2018-12-07 12:52:00 2023-10-26 15:06:00
9   301014561  Took tatawook 2017-08-10 14:22:03 2023-10-26 16:55:00
10  301015331  Wsky woversky 2017-02-09 11:21:52 2023-10-26 13:45:18
11  301018129  Ka.ş kiva a.ş 2018-09-15 15:53:45 2018-09-15 15:53:45
12  301019658  Bant boaurant 2017-05-26 17:37:46 2023-09-06 20:44:00
13  301020464  Asua axaxesua 2022-

## Müşterileri SES verisine Göre Kişiselleştir.
#### Verileri Endeksadan Çekeceğiz.

In [403]:
df_in_ses = pd.read_parquet('odeal_hackathon.parquet')

In [404]:
df_in_ses

id          Marka     IsyeriTipi UyeDurum  \
0       301002470  Kzar kü pazar   Tüzel Şirket    Aktif   
1       301002470  Kzar kü pazar   Tüzel Şirket    Aktif   
2       301002470  Kzar kü pazar   Tüzel Şirket    Aktif   
3       301002470  Kzar kü pazar   Tüzel Şirket    Aktif   
4       301002470  Kzar kü pazar   Tüzel Şirket    Aktif   
...           ...            ...            ...      ...   
999995  301033049  Öket özmarket  Şahıs Şirketi    Pasif   
999996  301033049  Öket özmarket  Şahıs Şirketi    Pasif   
999997  301033049  Öket özmarket  Şahıs Şirketi    Pasif   
999998  301033049  Öket özmarket  Şahıs Şirketi    Pasif   
999999  301033049  Öket özmarket  Şahıs Şirketi    Pasif   

                                                 UyeAdres  \
0                KALE MAH. GAZİ CAD. 82 C İLKADIM/ SAMSUN   
1                KALE MAH. GAZİ CAD. 82 C İLKADIM/ SAMSUN   
2                KALE MAH. GAZİ CAD. 82 C İLKADIM/ SAMSUN   
3                KALE MAH. GAZİ CAD. 82 C İLKADIM/ SAMSUN   
4                KALE MAH. GAZİ CAD. 82 C İLKADIM/ SAMSUN   
...                                                   ...   
999995  ZEYTİNLİ MAH. 79114 NOLU CAD. NO: 1 İÇ KAPI NO...   
999996  ZEYTİNLİ MAH. 79114 NOLU CAD. NO: 1 İÇ KAPI NO...   
999997  ZEYTİNLİ MAH. 79114 NOLU CAD. NO: 1 İÇ KAPI NO...   
999998  ZEYTİNLİ MAH. 79114 NOLU CAD. NO: 1 İÇ KAPI NO...   
999999  ZEYTİNLİ MAH. 79114 NOLU CAD. NO: 1 İÇ KAPI NO...   

                       Sehir   UyeAktivasyonTarih    IslemID  \
0             Samsun/Ilkadim  2015-09-08 15:08:23  401003136   
1             Samsun/Ilkadim  2015-09-08 15:08:23  403526702   
2             Samsun/Ilkadim  2015-09-08 15:08:23  403529006   
3             Samsun/Ilkadim  2015-09-08 15:08:23  403530803   
4             Samsun/Ilkadim  2015-09-08 15:08:23  403531107   
...                      ...                  ...        ...   
999995  Gaziantep/Şehitkamil  2019-01-31 15:47:21  547156206   
999996  Gaziantep/Şehitkamil  2019-01-31 15:47:21  547160265   
999997  Gaziantep/Şehitkamil  2019-01-31 15:47:21  547168209   
999998  Gaziantep/Şehitkamil  2019-01-31 15:47:21  547172655   
999999  Gaziantep/Şehitkamil  2019-01-31 15:47:21  547176522   

                 IslemTarih  IslemTutar    Tercih Versiyon SurumTarih  \
0       2015-09-18 14:35:04        5.00  CeptePos     None       None   
1       2019-04-13 13:31:00       32.70       ÖKC     None       None   
2       2019-04-13 19:30:00      998.05       ÖKC     None       None   
3       2019-04-14 13:52:00      105.75       ÖKC     None       None   
4       2019-04-14 15:09:00      163.00       ÖKC     None       None   
...                     ...         ...       ...      ...        ...   
999995  2023-06-09 21:35:00       16.50       ÖKC     None       None   
999996  2023-06-09 21:44:00       77.00       ÖKC     None       None   
999997  2023-06-09 22:01:00       37.00       ÖKC     None       None   
999998  2023-06-09 22:15:00       80.00       ÖKC     None       None   
999999  2023-06-09 22:21:00       18.00       ÖKC     None       None   

        Genel Nüfus   Nüfus Yoğunluğu  
0        8.167 Kişi  371,23 Kişi/km^2  
1        8.167 Kişi  371,23 Kişi/km^2  
2        8.167 Kişi  371,23 Kişi/km^2  
3        8.167 Kişi  371,23 Kişi/km^2  
4        8.167 Kişi  371,23 Kişi/km^2  
...             ...               ...  
999995  14.425 Kişi  133,56 Kişi/km^2  
999996  14.425 Kişi  133,56 Kişi/km^2  
999997  14.425 Kişi  133,56 Kişi/km^2  
999998  14.425 Kişi  133,56 Kişi/km^2  
999999  14.425 Kişi  133,56 Kişi/km^2  

[1000000 rows x 15 columns]

In [405]:
df_in_ses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 15 columns):
 #   Column              Non-Null Count    Dtype  
---  ------              --------------    -----  
 0   id                  1000000 non-null  int64  
 1   Marka               1000000 non-null  object 
 2   IsyeriTipi          1000000 non-null  object 
 3   UyeDurum            1000000 non-null  object 
 4   UyeAdres            1000000 non-null  object 
 5   Sehir               1000000 non-null  object 
 6   UyeAktivasyonTarih  1000000 non-null  object 
 7   IslemID             1000000 non-null  int64  
 8   IslemTarih          1000000 non-null  object 
 9   IslemTutar          1000000 non-null  float64
 10  Tercih              1000000 non-null  object 
 11  Versiyon            23643 non-null    object 
 12  SurumTarih          23643 non-null    object 
 13  Genel Nüfus         1000000 non-null  object 
 14  Nüfus Yoğunluğu     1000000 non-null  object 
dtypes: float64(1), i

In [406]:
df_in_ses.head().T

0  \
id                                                 301002470   
Marka                                          Kzar kü pazar   
IsyeriTipi                                      Tüzel Şirket   
UyeDurum                                               Aktif   
UyeAdres            KALE MAH. GAZİ CAD. 82 C İLKADIM/ SAMSUN   
Sehir                                         Samsun/Ilkadim   
UyeAktivasyonTarih                       2015-09-08 15:08:23   
IslemID                                            401003136   
IslemTarih                               2015-09-18 14:35:04   
IslemTutar                                               5.0   
Tercih                                              CeptePos   
Versiyon                                                None   
SurumTarih                                              None   
Genel Nüfus                                       8.167 Kişi   
Nüfus Yoğunluğu                             371,23 Kişi/km^2   

                                                           1  \
id                                                 301002470   
Marka                                          Kzar kü pazar   
IsyeriTipi                                      Tüzel Şirket   
UyeDurum                                               Aktif   
UyeAdres            KALE MAH. GAZİ CAD. 82 C İLKADIM/ SAMSUN   
Sehir                                         Samsun/Ilkadim   
UyeAktivasyonTarih                       2015-09-08 15:08:23   
IslemID                                            403526702   
IslemTarih                               2019-04-13 13:31:00   
IslemTutar                                              32.7   
Tercih                                                   ÖKC   
Versiyon                                                None   
SurumTarih                                              None   
Genel Nüfus                                       8.167 Kişi   
Nüfus Yoğunluğu                             371,23 Kişi/km^2   

                                                           2  \
id                                                 301002470   
Marka                                          Kzar kü pazar   
IsyeriTipi                                      Tüzel Şirket   
UyeDurum                                               Aktif   
UyeAdres            KALE MAH. GAZİ CAD. 82 C İLKADIM/ SAMSUN   
Sehir                                         Samsun/Ilkadim   
UyeAktivasyonTarih                       2015-09-08 15:08:23   
IslemID                                            403529006   
IslemTarih                               2019-04-13 19:30:00   
IslemTutar                                            998.05   
Tercih                                                   ÖKC   
Versiyon                                                None   
SurumTarih                                              None   
Genel Nüfus                                       8.167 Kişi   
Nüfus Yoğunluğu                             371,23 Kişi/km^2   

                                                           3  \
id                                                 301002470   
Marka                                          Kzar kü pazar   
IsyeriTipi                                      Tüzel Şirket   
UyeDurum                                               Aktif   
UyeAdres            KALE MAH. GAZİ CAD. 82 C İLKADIM/ SAMSUN   
Sehir                                         Samsun/Ilkadim   
UyeAktivasyonTarih                       2015-09-08 15:08:23   
IslemID                                            403530803   
IslemTarih                               2019-04-14 13:52:00   
IslemTutar                                            105.75   
Tercih                                                   ÖKC   
Versiyon                                                None   
SurumTarih                                              None   
Genel Nüfus                                       8.167 Kişi   
Nüfus Yoğunluğu         

In [407]:
with open('demografi_data.json', 'r') as file:
    demografi_data = json.load(file)

In [408]:
demografi_data

{'Samsun/Ilkadim': {'YASLI/GENC ORANI': {'Genc': 50.0,
   'Orta': 21.0,
   'Yasli': 29.0},
  'KADIN/ERKEK ORANI': {'Kadin': 49.0, 'Erkek': 51.0},
  'YASA GORE NUFUS DAGILIMI': {'0-4': 451,
   '5-9': 597,
   '10-14': 497,
   '15-19': 576,
   '20-24': 552,
   '25-29': 627,
   '30-34': 621,
   '35-39': 500,
   '40-44': 576,
   '45-49': 595,
   '50-54': 595,
   '55-59': 595,
   '60-64': 595,
   '65+': 1165},
  'MEDENI HAL': {'Evli': 3910,
   'Bekar': 1830,
   'Bosanmis': 520,
   'Eşi Ölmüş': 572},
  'EVLILIK ORANI': {'Evli': 37.0,
   'Bekar': 22.65,
   'Bosanmis': 47.88,
   'Eşi Ölmüş': 67.0},
  'EĞITIM DÜZEYI': {'Bilinmiyor': 86,
   'Okuryazar Değil': 159,
   'Okuryazar': 543,
   'İlkokul': 1858,
   'Ortaokul': 1243,
   'İlköğretim': 551,
   'Lise': 2011,
   'Lisans': 1140,
   'Yüksek Lisans': 76},
  'EĞITIM ORANI': {'Üniversite ve Üstü': 10.0,
   'İlk, Orta, Lise': 16.0,
   'Diğer': 72.0}},
 'Istanbul/Beşiktaş': {'YAŞLI/GENÇ ORANI': {'Genç': 19.0,
   'Orta': 22.0,
   'Yaşlı': 59.0},
  'K

In [410]:
df_in_ses['Genc_Orani'] = df_in_ses['Sehir'].map(lambda city: demografi_data.get(city, {}).get("YASLI/GENC ORANI", {}).get("Genc", 0))
df_in_ses['Orta_Orani'] = df_in_ses['Sehir'].map(lambda city: demografi_data.get(city, {}).get("YASLI/GENC ORANI", {}).get("Orta", 0))
df_in_ses['Yasli_Orani'] = df_in_ses['Sehir'].map(lambda city: demografi_data.get(city, {}).get("YASLI/GENC ORANI", {}).get("Yasli", 0))

df_in_ses['Kadin_Orani'] = df_in_ses['Sehir'].map(lambda city: demografi_data.get(city, {}).get("KADIN/ERKEK ORANI", {}).get("Kadin", 0))
df_in_ses['Erkek_Orani'] = df_in_ses['Sehir'].map(lambda city: demografi_data.get(city, {}).get("KADIN/ERKEK ORANI", {}).get("Erkek", 0))

df_in_ses['Universite_Ustu_Orani'] = df_in_ses['Sehir'].map(lambda city: demografi_data.get(city, {}).get("EĞITIM ORANI", {}).get("Üniversite ve Üstü", 0))
df_in_ses['Ilk_Orta_Lise_Orani'] = df_in_ses['Sehir'].map(lambda city: demografi_data.get(city, {}).get("EĞITIM ORANI", {}).get("İlk, Orta, Lise", 0))
df_in_ses['Diger_Egitim_Orani'] = df_in_ses['Sehir'].map(lambda city: demografi_data.get(city, {}).get("EĞITIM ORANI", {}).get("Diğer", 0))


In [411]:
df_in_ses

id          Marka     IsyeriTipi UyeDurum  \
0       301002470  Kzar kü pazar   Tüzel Şirket    Aktif   
1       301002470  Kzar kü pazar   Tüzel Şirket    Aktif   
2       301002470  Kzar kü pazar   Tüzel Şirket    Aktif   
3       301002470  Kzar kü pazar   Tüzel Şirket    Aktif   
4       301002470  Kzar kü pazar   Tüzel Şirket    Aktif   
...           ...            ...            ...      ...   
999995  301033049  Öket özmarket  Şahıs Şirketi    Pasif   
999996  301033049  Öket özmarket  Şahıs Şirketi    Pasif   
999997  301033049  Öket özmarket  Şahıs Şirketi    Pasif   
999998  301033049  Öket özmarket  Şahıs Şirketi    Pasif   
999999  301033049  Öket özmarket  Şahıs Şirketi    Pasif   

                                                 UyeAdres  \
0                KALE MAH. GAZİ CAD. 82 C İLKADIM/ SAMSUN   
1                KALE MAH. GAZİ CAD. 82 C İLKADIM/ SAMSUN   
2                KALE MAH. GAZİ CAD. 82 C İLKADIM/ SAMSUN   
3                KALE MAH. GAZİ CAD. 82 C İLKADIM/ SAMSUN   
4                KALE MAH. GAZİ CAD. 82 C İLKADIM/ SAMSUN   
...                                                   ...   
999995  ZEYTİNLİ MAH. 79114 NOLU CAD. NO: 1 İÇ KAPI NO...   
999996  ZEYTİNLİ MAH. 79114 NOLU CAD. NO: 1 İÇ KAPI NO...   
999997  ZEYTİNLİ MAH. 79114 NOLU CAD. NO: 1 İÇ KAPI NO...   
999998  ZEYTİNLİ MAH. 79114 NOLU CAD. NO: 1 İÇ KAPI NO...   
999999  ZEYTİNLİ MAH. 79114 NOLU CAD. NO: 1 İÇ KAPI NO...   

                       Sehir   UyeAktivasyonTarih    IslemID  \
0             Samsun/Ilkadim  2015-09-08 15:08:23  401003136   
1             Samsun/Ilkadim  2015-09-08 15:08:23  403526702   
2             Samsun/Ilkadim  2015-09-08 15:08:23  403529006   
3             Samsun/Ilkadim  2015-09-08 15:08:23  403530803   
4             Samsun/Ilkadim  2015-09-08 15:08:23  403531107   
...                      ...                  ...        ...   
999995  Gaziantep/Şehitkamil  2019-01-31 15:47:21  547156206   
999996  Gaziantep/Şehitkamil  2019-01-31 15:47:21  547160265   
999997  Gaziantep/Şehitkamil  2019-01-31 15:47:21  547168209   
999998  Gaziantep/Şehitkamil  2019-01-31 15:47:21  547172655   
999999  Gaziantep/Şehitkamil  2019-01-31 15:47:21  547176522   

                 IslemTarih  IslemTutar  ...  Genel Nüfus   Nüfus Yoğunluğu  \
0       2015-09-18 14:35:04        5.00  ...   8.167 Kişi  371,23 Kişi/km^2   
1       2019-04-13 13:31:00       32.70  ...   8.167 Kişi  371,23 Kişi/km^2   
2       2019-04-13 19:30:00      998.05  ...   8.167 Kişi  371,23 Kişi/km^2   
3       2019-04-14 13:52:00      105.75  ...   8.167 Kişi  371,23 Kişi/km^2   
4       2019-04-14 15:09:00      163.00  ...   8.167 Kişi  371,23 Kişi/km^2   
...                     ...         ...  ...          ...               ...   
999995  2023-06-09 21:35:00       16.50  ...  14.425 Kişi  133,56 Kişi/km^2   
999996  2023-06-09 21:44:00       77.00  ...  14.425 Kişi  133,56 Kişi/km^2   
999997  2023-06-09 22:01:00       37.00  ...  14.425 Kişi  133,56 Kişi/km^2   
999998  2023-06-09 22:15:00       80.00  ...  14.425 Kişi  133,56 Kişi/km^2   
999999  2023-06-09 22:21:00       18.00  ...  14.425 Kişi  133,56 Kişi/km^2   

       Genc_Orani Orta_Orani Yasli_Orani  Kadin_Orani Erkek_Orani  \
0            50.0       21.0        29.0         49.0        51.0   
1            50.0       21.0        29.0         49.0        51.0   
2            50.0       21.0        29.0         49.0        51.0   
3            50.0       21.0        29.0         49.0        51.0   
4            50.0       21.0        29.0         49.0        51.0   
...           ...        ...         ...          ...         ...   
999995      945.0      649.0        96.0         50.0        50.0   
999996      945.0      649.0        96.0         50.0        50.0   
999997      945.0      649.0        96.0         50.0        50.0   
999998      945.0      649.0        96.0         50.0        50.0   
999999      945.0      649.0        96.0         50.0        50.0   

        Univer

#### Yeni veriler ile birlikte 23 kolonlu bir veri seti oldu.
Data columns (total 23 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   id                     1000000 non-null  int64  
 1   Marka                  1000000 non-null  object 
 2   IsyeriTipi             1000000 non-null  object 
 3   UyeDurum               1000000 non-null  object 
 4   UyeAdres               1000000 non-null  object 
 5   Sehir                  1000000 non-null  object 
 6   UyeAktivasyonTarih     1000000 non-null  object 
 7   IslemID                1000000 non-null  int64  
 8   IslemTarih             1000000 non-null  object 
 9   IslemTutar             1000000 non-null  float64
 10  Tercih                 1000000 non-null  object 
 11  Versiyon               23643 non-null    object 
 12  SurumTarih             23643 non-null    object 
 13  Genel Nüfus            1000000 non-null  object 
 14  Nüfus Yoğunluğu        1000000 non-null  object 
 15  Genc_Orani             1000000 non-null  float64
 16  Orta_Orani             1000000 non-null  object 
 17  Yasli_Orani            1000000 non-null  float64
 18  Kadin_Orani            1000000 non-null  float64
 19  Erkek_Orani            1000000 non-null  float64
 20  Universite_Ustu_Orani  1000000 non-null  float64
 21  Ilk_Orta_Lise_Orani    1000000 non-null  float64
 22  Diger_Egitim_Orani     1000000 non-null  float64

### 1. aktif müşteriyi ödüllendir.
Bu müşteri 2015'ten beri aktif ve belirli aralıklarla alışveriş yapmış. 
Sadakat programları veya indirimlerle bu tür müşterileri ödüllendirebiliriz vs bişiler.

In [419]:
grouped['duration'] = grouped['max'] - grouped['min']
sorted_df = grouped.sort_values(by='duration', ascending=False)
print(sorted_df.head(2))  # Alışveriş süresi en fazla olan ilk 2 müşteri


          id          Marka                 min                 max  \
0  301002470  Kzar kü pazar 2015-09-18 14:35:04 2023-10-26 10:47:00   
6  301009682  Wtma waarıtma 2016-05-19 10:29:01 2023-10-26 16:39:00   

            duration  
0 2959 days 20:11:56  
6 2716 days 06:09:59  


### 2. Uzun Zamandır Alışveriş Yapmayanları Geri Kazan Kampanyası Olabilir..

In [422]:
last_purchase = df_in_ses.groupby('id')['IslemTarih'].max().reset_index()

last_purchase['days_since_last_purchase'] = (datetime.now() - last_purchase['IslemTarih']).dt.days

# 180 günden daha uzun süre alışveriş yapmamış müşteriler
inactive_customers = last_purchase[last_purchase['days_since_last_purchase'] > 180].sort_values(by='days_since_last_purchase', ascending=False)

inactive_customers

id          IslemTarih  days_since_last_purchase
19  301025251 2018-07-24 11:03:37                      1924
11  301018129 2018-09-15 15:53:45                      1870
30  301029860 2019-11-28 14:26:47                      1431
7   301010389 2020-05-10 11:09:14                      1268
37  301032496 2020-11-25 22:10:00                      1068
1   301002583 2020-12-23 12:28:28                      1041
25  301026648 2021-04-30 17:08:00                       912
27  301026958 2021-05-04 21:19:38                       908
33  301031149 2021-05-19 11:54:00                       894
3   301006906 2022-09-12 18:30:00                       412
4   301009366 2022-09-19 18:01:00                       405
29  301029464 2022-11-14 11:07:00                       350

### 3. Az Alışverişi olan Markaları Teşvik Et.
#### 3.1 Alışveriş sıklığına göre(frekans alcaz):


In [426]:
az_alisveris_yapilan_markalar = df_in_ses.groupby('Marka').size().reset_index(name='Total_Transactions').sort_values(by='Total_Transactions').head(5)

az_alisveris_yapilan_markalar


Marka  Total_Transactions
21  Ka.ş kiva a.ş                   2
19  Hkek haan kek                   4
26  Llım leazılım                   5
14    Ezgi ezezgi                  10
32  Teri tejileri                 143

#### 3.2 Toplam Alışveriş Tutarına Göre:

In [427]:
# Markalara göre toplam alışveriş tutarı
az_tutarli_markalar = df_in_ses.groupby('Marka')['IslemTutar'].sum().reset_index().sort_values(by='IslemTutar').head(10)

print(az_tutarli_markalar)

            Marka  IslemTutar
21  Ka.ş kiva a.ş       20.00
14    Ezgi ezezgi       35.25
26  Llım leazılım      500.00
19  Hkek haan kek     1228.17
32  Teri tejileri     2044.92
38  Örek öz börek   451458.73
4   Ayon arvasyon   538330.90
2   Alik arncilik   558967.49
22  Kayi kol bayi   767724.72
27  Mand mad land   804918.38


#### Belki 3.1 ve 3.2 yi harmanlayan bi yaklasım gelistiririz. Yeni bir algoritma ..

### 4. Ödeal ın şikayet var gibi yerlerde insanların şikayetlerine bakarak genel kategorisini genişlet.

# 5 Yeni Üyelik Kampanyaları yap.

In [435]:
en_son_uye_tarihi = df_in_ses['UyeAktivasyonTarih'].max()

yeni_uyeler = df_in_ses[df_in_ses['UyeAktivasyonTarih'] >= en_son_uye_tarihi]

print(yeni_uyeler.head(5))


               id          Marka    IsyeriTipi UyeDurum  \
300975  301024729  Aüfe ass büfe  Tüzel Şirket    Pasif   
300976  301024729  Aüfe ass büfe  Tüzel Şirket    Pasif   
300977  301024729  Aüfe ass büfe  Tüzel Şirket    Pasif   
300978  301024729  Aüfe ass büfe  Tüzel Şirket    Pasif   
300979  301024729  Aüfe ass büfe  Tüzel Şirket    Pasif   

                                                 UyeAdres              Sehir  \
300975  MEHMET NESİH ÖZMEN MAH. KESTANE SK. AS CAFE NO...  Istanbul/Güngören   
300976  MEHMET NESİH ÖZMEN MAH. KESTANE SK. AS CAFE NO...  Istanbul/Güngören   
300977  MEHMET NESİH ÖZMEN MAH. KESTANE SK. AS CAFE NO...  Istanbul/Güngören   
300978  MEHMET NESİH ÖZMEN MAH. KESTANE SK. AS CAFE NO...  Istanbul/Güngören   
300979  MEHMET NESİH ÖZMEN MAH. KESTANE SK. AS CAFE NO...  Istanbul/Güngören   

         UyeAktivasyonTarih    IslemID          IslemTarih  IslemTutar  ...  \
300975  2022-12-10 15:38:46  403674129 2019-05-17 16:30:00         2.5  ...   
300976

#### En son üye olan adamı id den bulabilir miyim bunu anlamak için en son id adresini alıp bunun UyeAktivasyonTarihlerine baktım. 1.ayda aktif üyelik yapılmış dolayısıyla id den belli olmuyor bu işler..

In [448]:
son_id = max(df_in_ses['id'])
print(f'Son IDli üye: {son_id}.\Bu üyenin aktivasyon tarihi:')
a = df_in_ses[df_in_ses['id']==son_id].UyeAktivasyonTarih
a.values[0]

Son IDli üye: 301033049.\Bu üyenin aktivasyon tarihi:


'2019-01-31 15:47:21'

# Kampanya Önerileri

### Yerelleşme Stratejisi:

Sehir özelliği kullanılarak belirli bir şehirdeki tüketicilere özel kampanyalar tasarlanabilir. Örneğin: İstanbul'daki tüketicilere özel bir indirim veya Ankara'daki tüketicilere yönelik özel bir ürün tanıtımı.

### Yaş Grubu Odaklı Kampanyalar:

Genc_Orani, Orta_Orani ve Yasli_Orani özelliklerine dayanarak belirli yaş gruplarına yönelik kampanyalar oluşturabilirsiniz. Örneğin: Genç nüfusu yüksek bir şehirde gençlere yönelik moda veya teknoloji ürünleri tanıtımı. Ya da yaşlı nüfusu yüksek bölgelerde sağlık ve konfora yönelik ürünlerin tanıtımı.

### Cinsiyet Odaklı Kampanyalar:

Kadin_Orani ve Erkek_Orani özellikleri kullanarak cinsiyete özel kampanyalar düzenlenebilir. Örneğin: Kadın nüfusu yüksek bölgelerde kadınlara özel kozmetik ürün tanıtımları veya erkek nüfusu yüksek bölgelerde erkeklere özel hobi ürünleri tanıtımları.

### Eğitim Seviyesine Göre Ürün Tanıtımları:

Universite_Ustu_Orani, Ilk_Orta_Lise_Orani ve Diger_Egitim_Orani özellikleriyle eğitim seviyesine göre kampanyalar oluşturabilirsiniz. Örneğin: Üniversite ve üzeri eğitim oranı yüksek bölgelerde bilimsel kitaplar veya profesyonel eğitim materyallerinin tanıtımı.

### Yeni Üye Kampanyaları:

UyeAktivasyonTarih özelliği kullanılarak son dönemde üye olanlara özel indirimler veya hoş geldin paketleri sunabilirsiniz.

### Özel Tercihlere Göre Kampanyalar:

Tercih özelliğiyle, kullanıcıların belirli bir ürün veya hizmete olan ilgisine göre kişiselleştirilmiş kampanyalar oluşturabilirsiniz.

### Versiyon Güncellemeleri:

Versiyon ve SurumTarih özellikleriyle, belirli bir ürünün yeni sürümünü tanıtarak, eski sürümü kullananlara özel indirimler sunabilirsiniz.